smartsim/control/manifest.py

```diff
@@ -0,0 +1,102 @@
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from ..database import Orchestrator
+from ..entity import EntityList, SmartSimEntity
+from ..error import SmartSimError
+from ..error.errors import SmartSimError
+
+# List of types derived from EntityList which require specific behavior
+# A corresponding property needs to exist (like db for Orchestrator),
+# otherwise they will not be accessible
+entity_list_exception_types = [Orchestrator]
+
+
+class Manifest:
+    """This class is used to keep track of all deployables generated by an experiment.
+    Different types of deployables (i.e. different `SmartSimEntity`-derived objects)
+    can be accessed by using the corresponding accessor.
+
+    Instances of ``Model``, ``Ensemble`` and ``Orchestrator``
+    can all be passed as arguments
+    """
+
+    def __init__(self, *args):
+        self._deployables = list(args)
+        self._check_types(self._deployables)
+        self._check_names(self._deployables)
+
+    @property
+    def db(self):
+        _db = None
+        for deployable in self._deployables:
+            if isinstance(deployable, Orchestrator):
+                if _db:
+                    raise SmartSimError("Manifest was given two orchestrators")
+                _db = deployable
+        return _db
+
+    @property
+    def models(self):
+        _models = []
+        for deployable in self._deployables:
+            if isinstance(deployable, SmartSimEntity):
+                _models.append(deployable)
+        return _models
+
+    @property
+    def ensembles(self):
+        _ensembles = []
+        for deployable in self._deployables:
+            if isinstance(deployable, EntityList):
+                is_exceptional_type = False
+                for exceptional_type in entity_list_exception_types:
+                    if isinstance(deployable, exceptional_type):
+                        is_exceptional_type |= True
+                if not is_exceptional_type:
+                    _ensembles.append(deployable)
+
+        return _ensembles
+
+    def _check_names(self, deployables):
+        used = []
+        for deployable in deployables:
+            name = getattr(deployable, "name", None)
+            if not name:
+                raise AttributeError(f"Entity has no name. Please set name attribute.")
+            if name in used:
+                raise SmartSimError("User provided two entities with the same name")
```

Comment:
When I refactor the jobManager with the different classes for jobs, we should put in a check where the manifest is passed to the JM to see if any new entities are defined by the same name

Reply:
OK, it makes sense to me.